bus-mapping/src/evm/opcodes/calldataload.rs

```diff
@@ -0,0 +1,207 @@
+use crate::{
+    circuit_input_builder::{CircuitInputStateRef, ExecStep},
+    operation::{CallContextField, CallContextOp, RW},
+    Error,
+};
+use eth_types::{GethExecStep, U256};
+
+use super::Opcode;
+
+#[derive(Clone, Copy, Debug)]
+pub(crate) struct Calldataload;
+
+impl Opcode for Calldataload {
+    fn gen_associated_ops(
+        state: &mut CircuitInputStateRef,
+        geth_steps: &[GethExecStep],
+    ) -> Result<Vec<ExecStep>, Error> {
+        let geth_step = &geth_steps[0];
+        let mut exec_step = state.new_step(geth_step)?;
+
+        // fetch the top of the stack, i.e. offset in calldata to start reading 32-bytes
+        // from.
+        let offset = geth_step.stack.nth_last(0)?;
+
+        state.push_stack_op(
+            &mut exec_step,
+            RW::READ,
+            geth_step.stack.nth_last_filled(0),
+            offset,
+        )?;
+
+        state.push_op(
+            &mut exec_step,
+            RW::READ,
+            CallContextOp {
+                call_id: state.call()?.call_id,
+                field: CallContextField::TxId,
+                value: state.tx_ctx.id().into(),
+            },
+        );
+
+        if !state.call()?.is_root {
+            state.push_op(
+                &mut exec_step,
+                RW::READ,
+                CallContextOp {
+                    call_id: state.call()?.call_id,
+                    field: CallContextField::CallDataLength,
+                    value: state.call()?.call_data_length.into(),
+                },
+            );
+            state.push_op(
+                &mut exec_step,
+                RW::READ,
+                CallContextOp {
+                    call_id: state.call()?.call_id,
+                    field: CallContextField::CallDataOffset,
+                    value: state.call()?.call_data_offset.into(),
+                },
+            );
+            state.push_op(
+                &mut exec_step,
+                RW::READ,
+                CallContextOp {
+                    call_id: state.call()?.call_id,
+                    field: CallContextField::CallerId,
+                    value: state.call()?.caller_id.into(),
+                },
+            );
+        }
+
+        let (offset, src_addr_end) = (
+            offset.as_usize(),
+            state.call()?.call_data_offset as usize + state.call()?.call_data_length as usize,
+        );
+        let is_root = state.call()?.is_root;
+        let calldata_word = (0..32)
+            .map(|idx| {
+                let addr = offset + idx;
+                if addr < src_addr_end {
+                    if is_root {
+                        state.tx.input[addr]
+                    } else {
+                        // TODO: read caller memory
+                        unimplemented!();
```

Comment:
Why isn't this implemented? Is there anything that makes the implementation of the caller memory bytes to be read?

Reply:
I see!

Please, fix `CALLDATACOPY` in a sepparated PR please.